File: gams/apifiles/C/xp_example1.c
/*
  Use this command to compile the example:
  cl xp_example1.c api/gdxcc.c -Iapi
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "gclgms.h"
#include "gdxcc.h"

static gdxStrIndexPtrs_t Indx;
static gdxStrIndex_t     IndxXXX; 
static gdxValues_t       Values;

void ReportGDXError(gdxHandle_t PGX) {
  char S[GMS_SSSIZE];

  printf("**** Fatal GDX Error\n");
  gdxErrorStr(PGX, gdxGetLastError(PGX), S);
  printf("**** %s\n", S);
  exit(1);
}

void ReportIOError(int N, const char *msg ) {
  printf("**** Fatal I/O Error = %d when calling %s\n",N,msg);
  exit(1);
}

void WriteData(gdxHandle_t PGX, const char *s, const double V) {
  strcpy(Indx[0],s);
  Values[GMS_VAL_LEVEL] = V;
  gdxDataWriteStr(PGX,Indx,Values);
}

int main (int argc, char *argv[]) {

  gdxHandle_t PGX=NULL;
  char        Msg[GMS_SSSIZE];
  char        Producer[GMS_SSSIZE];
  char        Sysdir[GMS_SSSIZE];
  int         ErrNr;
  int         VarNr;
  int         NrRecs;
  int         N;
  int         Dim;
  char        VarName[GMS_SSSIZE];
  int         VarTyp;
  int         D;
  
  if (argc < 2 || argc > 3) {
    printf("**** xp_Example1: incorrect number of parameters\n");
    exit(1);
  }

  if(strlen(argv[1]) >= GMS_SSSIZE) {
    printf("*** Your system directory (argument 1) cannot exceed 255 characters\n");
    exit(1);
  }
  
  strcpy(Sysdir, argv[1]);
  printf("xp_Example1 using GAMS system directory: %s\n", Sysdir);

  if (!gdxCreateD(&PGX, Sysdir, Msg, sizeof(Msg))) {
    printf("**** Could not load GDX library\n");
    printf("**** %s\n", Msg);
    exit(1);
  }

  gdxGetDLLVersion(PGX, Msg);
  printf("Using GDX DLL version: %s\n", Msg);

  GDXSTRINDEXPTRS_INIT(IndxXXX,Indx);

  if (2 == argc) {
    /* Write demand data */
    gdxOpenWrite(PGX, "demanddata.gdx", "xp_example1", &ErrNr);
    if (ErrNr) ReportIOError(ErrNr,"gdxOpenWrite");
    if (!gdxDataWriteStrStart(PGX,"Demand","Demand data",1,GMS_DT_PAR ,0))
      ReportGDXError(PGX);
    WriteData(PGX,"New-York",324.0);
    WriteData(PGX,"Chicago" ,299.0);
    WriteData(PGX,"Topeka"  ,274.0);
    if (!gdxDataWriteDone(PGX)) ReportGDXError(PGX);
    printf("Demand data written by xp_example1\n");
  } else {
    gdxOpenRead(PGX, argv[2], &ErrNr);
    if (ErrNr) ReportIOError(ErrNr,"gdxOpenRead");
    gdxFileVersion(PGX,Msg,Producer);
    printf("GDX file written using version: %s\n",Msg);
    printf("GDX file written by: %s\n",Producer);

    if (!gdxFindSymbol(PGX,"x",&VarNr)) {
      printf("**** Could not find variable X\n");
      exit(1);
    }

    gdxSymbolInfo(PGX,VarNr,VarName,&Dim,&VarTyp);
    if (Dim != 2 || GMS_DT_VAR != VarTyp) {
      printf("**** X is not a two dimensional variable: %d:%d\n",Dim,VarTyp);
      exit(1);
    }

    if (!gdxDataReadStrStart(PGX,VarNr,&NrRecs)) ReportGDXError(PGX);
    printf("Variable X has %d records\n", NrRecs);
    while (gdxDataReadStr(PGX,Indx,Values,&N)) {
      if (0 == Values[GMS_VAL_LEVEL]) continue; /* skip level 0.0 is default */
      for (D=0; D<Dim; D++) printf("%c%s", D? '.':' ', Indx[D]);
      printf(" = %7.2f\n", Values[GMS_VAL_LEVEL]);
    }
    printf("All solution values shown\n");
    gdxDataReadDone(PGX);
  }

  if (ErrNr = gdxClose(PGX)) ReportIOError(ErrNr,"gdxClose");

  if (!gdxFree(&PGX)) {
    printf("Problems unloading the GDX DLL\n");
    exit(1);
  }

  return 0;
} /* main */

File: gams/apifiles/CSharp/xp_example1.cs
///////////////////////////////////////////////////////////////
// This program generates demand data for a modified version //
// of the trnsport model or reads the solution back from a   //
// gdx file.                                                 //
//                                                           //
// Calling convention:                                       //
// Case 1:                                                   //
//    Parameter 1: GAMS system directory                     //
// The program creates a GDX file with demand data           //
// Case 2:                                                   //
//    Parameter 1: GAMS system directory                     //
//    Parameter 2: gdxfile                                   //
// The program reads the solution from the GDX file          //
// Paul van der Eijk Jun-12, 2002                            //
///////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace xp_example1
{
    class xp_example1
    {
        static gdxcs gdx;
        static void ReportGDXError()
        {
            string S = string.Empty;
            Console.WriteLine("**** Fatal GDX Error");
            gdx.gdxErrorStr(gdx.gdxGetLastError(), ref S);
            Console.WriteLine("**** " + S);
            Environment.Exit(1);
        }

        static void ReportIOError(int N)
        {
            Console.WriteLine("**** Fatal I/O Error = " + N);
            Environment.Exit(1);
        }

        static void WriteData(string s, double V)
        {
            string[] Indx = new string[gamsglobals.maxdim];
            double[] Values = new double[gamsglobals.val_max];
            Indx[0] = s;
            Values[gamsglobals.val_level] = V;
            gdx.gdxDataWriteStr(Indx, Values);
        }

        static int Main(string[] args)
        {
            string Msg = string.Empty;
            string Sysdir;
            string Producer = string.Empty;
            int ErrNr = 0;
            int rc;
            string[] Indx = new string[gamsglobals.maxdim];
            double[] Values = new double[gamsglobals.val_max];
            int VarNr = 0;
            int NrRecs = 0;
            int N = 0;
            int Dimen = 0;
            string VarName = string.Empty;
            int VarTyp = 0;
            int D;

            if (Environment.GetCommandLineArgs().Length != 2 && Environment.GetCommandLineArgs().Length != 3)
            {
                Console.WriteLine("**** XP_Example1: incorrect number of parameters");
                return 1;
            }

            String[] arguments = Environment.GetCommandLineArgs();
            Sysdir = arguments[1];
            Console.WriteLine("XP_Example1 using GAMS system directory: " + Sysdir);

            gdx = new gdxcs(Sysdir, ref Msg);
            if (Msg != string.Empty)
            {
                Console.WriteLine("**** Could not load GDX library");
                Console.WriteLine("**** " + Msg);
                return 1;
            }

            gdx.gdxGetDLLVersion(ref Msg);
            Console.WriteLine("Using GDX DLL version: " + Msg);

            if (Environment.GetCommandLineArgs().Length == 2)
            {
                //write demand data
                gdx.gdxOpenWrite("demanddata.gdx", "xp_example1", ref ErrNr);
                if (ErrNr != 0) xp_example1.ReportIOError(ErrNr);
                if (gdx.gdxDataWriteStrStart("Demand", "Demand data", 1, gamsglobals.dt_par, 0) == 0) ReportGDXError();
                WriteData("New-York", 324.0);
                WriteData("Chicago", 299.0);
                WriteData("Topeka", 274.0);
                if (gdx.gdxDataWriteDone() == 0) ReportGDXError();
                Console.WriteLine("Demand data written by xp_example1");
            }
            else
            {
                rc = gdx.gdxOpenRead(arguments[2], ref ErrNr);
                if (ErrNr != 0) ReportIOError(ErrNr);

                //read x variable back (non-default level values only)
                gdx.gdxFileVersion(ref Msg, ref Producer);
                Console.WriteLine("GDX file written using version: " + Msg);
                Console.WriteLine("GDX file written by: " + Producer);

                if (gdx.gdxFindSymbol("x", ref VarNr) == 0)
                {
                    Console.WriteLine("**** Could not find variable X");
                    return 1;
                }

                gdx.gdxSymbolInfo(VarNr, ref VarName, ref Dimen, ref VarTyp);
                if (Dimen != 2 || VarTyp != gamsglobals.dt_var)
                {
                    Console.WriteLine("**** X is not a two dimensional variable");
                    return 1;
                }

                if (gdx.gdxDataReadStrStart(VarNr, ref NrRecs) == 0) ReportGDXError();

                Console.WriteLine("Variable X has " + NrRecs + " records");
                while (gdx.gdxDataReadStr(ref Indx, ref Values, ref N) != 0)
                {
                    if(Values[gamsglobals.val_level] == 0.0) //skip level = 0.0 is default
                        continue;
                    for (D=0; D<Dimen; D++)
                    {
                        Console.Write(Indx[D]);
                        if (D < Dimen-1) Console.Write(".");
                    }
                    Console.WriteLine(" = " + Values[gamsglobals.val_level]);
                }
                Console.WriteLine("All solution values shown");
                gdx.gdxDataReadDone();
            }
            ErrNr = gdx.gdxClose();
            if (ErrNr != 0) ReportIOError(ErrNr);
            return 0;
        }
    }
}
